### DISCUSSION TITLE
Deprecating method with nullable parameters: ambiguous call when passing null argument

### DISCUSSION THREAD

**Discussion by eduherminio** [dotnet/roslyn#61848] [Q&A]
Given this method in a library, compiled with `<Nullable>Enable</Nullable>`:

```csharp
public static async Task Solve(SolverConfiguration? configuration, params uint[] problemNumbers) { }
```

I'm trying to deprecate it in favor of one that accepts `Action<SolverConfiguration>?` instead of the config object.

This is my attempt to doing it without breaking compatibility:

```csharp
[Obsolete("Trying to deprecate this: made it non-optional and non-nullable")]
public static async Task Solve(SolverConfiguration configuration, params uint[] problemNumbers) { }

public static async Task Solve(Action<SolverConfiguration>? options = null, params uint[] problemNumbers) { }
```

However, I still get the following error when trying to do `await Solver.Solve(null, 1, 2)` within a project compiled with `<Nullable>Enable</Nullable>`:

```
CS0121	The call is ambiguous between the following methods or properties: 'Solver.Solve(Action<SolverConfiguration>?, params uint[])' and 'Solver.Solve(SolverConfiguration, params uint[])'
```

![image](https://user-images.githubusercontent.com/11148519/173370172-95e53d70-03fd-4023-8395-53740abc582b.png)

Extended behavior:

```csharp
await Solver.Solve(null, 1, 2);                  // Fails to compile, no matter Nullable is enabled or not
await Solver.Solve(options: null, 1, 2);         // Compiles
await Solver.Solve(configuration: null, 1, 2);   // Compiles only in projects with Nullable disabled
```

My questions are:

* Why is `await Solver.Solve(null, 1, 2)` ambiguous in projects with `Nullable` enabled?

* Is there any way of refactoring my code so that I can deprecate one of the methods without possibly generating a breaking change, no matter the `Nullable` configuration of my library clients.?

**Comment by RikkiGibson** [comment]
It makes sense to me that there's an ambiguity here. `null` converts to both `Action<SolverConfiguration>` and `SolverConfiguration` (assuming it is a class type), and neither conversion is *better* per the language rules. The `<Nullable>` property doesn't affect this because the goal of the feature is to avoid affecting runtime semantics. Affecting overload resolution results essentially falls in that category.

Adding an argument label can fix the ambiguity because overloads where the parameter names don't match the labels can be ruled out.

> ```cs
> await Solver.Solve(configuration: null, 1, 2);   // Compiles only in projects with Nullable disabled
> ```
> 

This bit seems concerning. If you are able to reproduce this minimally in a small project or in a scratchpad like https://sharplab.io it would be great if you could file it as a bug.

**Reply by eduherminio** [reply]
The ambiguity makes sense in that case, thanks.
Am I missing any obvious way to make both methods coexist (one of them decorated with `[Obsolete]` attribute) while avoiding this ambiguity and therefore generating a breaking change for those ones that were invoking it like `await Solver.Solve(null, 1, 2)` ?


> This bit seems concerning. If you are able to reproduce this minimally in a small project or in a scratchpad like https://sharplab.io/ it would be great if you could file it as a bug.

My bad, as @PathogenDavid hinted I have CS8625 configured as error, so the behavior is as expected.

**Comment by PathogenDavid** [comment]
> * Why is `await Solver.Solve(null, 1, 2)` ambiguous in projects with `Nullable` enabled?

Overload resolution doesn't consider nullability. Like Rikki said, both `Action<SolverConfiguration>` and `SolverConfiguration` can both be null so it's ambiguous.

> ```csharp
> await Solver.Solve(configuration: null, 1, 2);   // Compiles only in projects with Nullable disabled
> ```

This should compile regardless, but it will be a warning with nullable enabled since you marked `configuration` as non-nullable. ([SharpLab](https://sharplab.io/#v2:EYLgtghglgdgNAFxANwKYCcoDMCeAfAYhgFcAbUiYU1AAlRkuoFgAoAgB3QgHNIaB3COhixuNACZQAzo1oBhAMoBGAJwqAHDQD0WmgAFN/KAgAWAe2IJ9K1hy68IAoSJhjJMqvIUAGAGxLNHRoAQQBjUNQpKVEaM2ApM2oEWjBUMGAMVj0AJiUs730lAFYAblYsgGZC331smgVEtHRWAG9WGg6aAG0AeXjE1GSACgAiABV0HBiEMwlUTlRQiGSaU2kQGkhxWmMaGDMYAFozdgQoA4hSGggYcT2Dw5JyWRGASgBdds69Kr0lGoM+hqDVIaCGIKacgOWCg3GIXDOBxooWhsPhy3O8Bo7CEEDAUhoxFgCC672x6Di1DAADliOkMFJXl9Op0ALwAPlWJgp/D2qF51LMCAAkmB2FT6MlxABRAAeEVOmKGrzKLGZ+l+/30mj0wMaqCGf2yAB4IRgoTAYXCEZj2QB+WKKg4E1l7MikODY3H4wnE0nkylpWn09CM9Usjlcnl8gVC0XitKS1Ay+XzREwZWqgC+5RYP1q9X16AtVvR6ZoLRoObVbCeFE8dAYDaCYzM3G41E2qFz+b+ALqY0iCFa6t7WsBupoAFlBuZxMr1W0WCyWTo9CpC6CMAA6M1DOuepSe7Iqldn89BABi0FIBJmyLMYqg1E9+02y2S6BotOeDekjdkO4zC/fZh2Xc9103Jpd31IYTnTKQNgPGgjxoE8SnPDogihJ9qCkcNvg3M10BgrchhRS00RtA4kPdQ9j1PLDdBw9hn0iWIYFIHAaFgAMACtFgQAkjFMb93VkCRpEA9VqyzIA==)) Did you configure CS8625 as an error?
